src/backend/distributed/planner/fast_path_router_planner.c

```diff
@@ -225,52 +278,69 @@ FastPathRouterQuery(Query *query, Node **distributionKeyValue)
 		return false;
 	}
 
+	bool isFastPath = false;
+	bool canAvoidDeparse = false;
+	Node *distributionKeyValue = NULL;
+
 	/*
 	 * If the table doesn't have a distribution column, we don't need to
 	 * check anything further.
 	 */
 	Var *distributionKey = PartitionColumn(distributedTableId, 1);
 	if (!distributionKey)
 	{
-		return true;
+		/* Local execution may avoid a deparse on single shard distributed tables */
+		canAvoidDeparse = IsCitusTableTypeCacheEntry(cacheEntry,
+													 SINGLE_SHARD_DISTRIBUTED) ||
+						  IsCitusTableTypeCacheEntry(cacheEntry, CITUS_LOCAL_TABLE);
```

Comment:
should we add such a comment here?

```
/*
 * We don't yet support reference tables in this code-path since
 * modifications on reference tables are complicated enough to support
 * here.
 */
```